### PR TITLE
OMTF configuration fix to use phase-2 values

### PR DESCRIPTION
**Pull request by folguera**
#### PR description:

This contains the required config files to load the correct OMTF configuration for Phase-2. This is built on top of #1241 and it requires https://github.com/cms-data/L1Trigger-L1TMuon/pull/27 to be merged and downloaded inside L1Trigger/L1TMuon/data/omtf_data 

#### PR validation:

First quick look at efficiencies look OK: https://folguera.web.cern.ch/L1T/2024_04_AR_Tests/V36nano_muons_test/object_performance/turnons/MuonTFsMatching_Eta_Pt15toInf_-999_V36nano_muons.png

### CODE REVIEW

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 2 files that did not meet formatting requirements:
 - L1Trigger/L1TMuon/src/MicroGMTConfiguration.cc
 - L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTFwdMuonTranslator.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

This PR passes available unit tests!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b runtests`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation

I found a non-zero return code running the relval workflows for this PR!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`1`|
|command|`eval `scramv1 runtime -sh` && runTheMatrix.py --what upgrade -l 26834.78`|

**Comment by folguera** [comment]
Hi @aloeliger @epalencia, this seems to be ready now, from a quick look, the efficiency for GMT SA muons is recovered (running with more stats now) I'd appreciate is @kbunkow and @zhenbinwu could have a look as well. This superseeds #1241

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 1 files that did not meet formatting requirements:
 - L1Trigger/L1TMuon/src/MicroGMTConfiguration.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

This PR passes available unit tests!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b runtests`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation

I found a non-zero return code running the relval workflows for this PR!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`1`|
|command|`eval `scramv1 runtime -sh` && runTheMatrix.py --what upgrade -l 26834.78`|

File: L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTFwdMuonTranslator.cc [hunk 1]
```diff
@@ -100,19 +100,19 @@ void Phase2L1TGMTFwdMuonTranslator::produce(edm::Event& iEvent, const edm::Event
...
-    associateStubs(samuon, stubs);
-
-    // Add To Collections
-    if (mu.hwPt() > 0)
+      //now associate the stubs
+      associateStubs(samuon, stubs);
       prompt.push_back(samuon);
-    else if (mu.hwPtUnconstrained() > 0)
+    }
+    if (mu.hwPtUnconstrained() > 0)  //Assume exculsive, need double check
```
Comment (zhenbinwu) [inline L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTFwdMuonTranslator.cc]: Can you remove the old comments?

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

Attempts to compile this PR succeeded!

|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -j 8`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found no issues with the code checks!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-checks && scram b -k -j 8 code-checks`|

I found no issues with the headers!
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 check-headers`|

**Comment by triggerDoctor** [comment]
Hello, I'm triggerDoctor. @aloeliger is testing this script for L1T offline software validation.

I found 2 files that did not meet formatting requirements:
 - L1Trigger/L1TMuon/src/MicroGMTConfiguration.cc
 - L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTFwdMuonTranslator.cc

Please run `scram b code-format` to auto-apply code formatting
|    Info   |     Value    |
|:---------:|:------------:|
|return code|`0`|
|command|`eval `scramv1 runtime -sh` && scram b -k -j 8 code-format-all`|

**Comment by zhenbinwu** [comment]
It doesn't work for me. I got the missing file error:
`edm::FileInPath unable to find file L1Trigger/L1TMuon/data/omtf_config/Patterns_ExtraplMB1nadMB2DTQualAndEtaFixedP_ValueP1Scale_t20_v1_SingleMu_iPt_and_OneOverPt_classProb17_recalib2_minDP0.xml anywhere in the search path.`

**Comment by zhenbinwu** [comment]
Unfortunately it is not working for me. It miss the xml file 
`edm::FileInPath unable to find file L1Trigger/L1TMuon/data/omtf_config/Patterns_ExtraplMB1nadMB2DTQualAndEtaFixedP_ValueP1Scale_t20_v1_SingleMu_iPt_and_OneOverPt_classProb17_recalib2_minDP0.xml anywhere in the search path.`

I can't find it from the cms-data : https://github.com/cms-data/L1Trigger-L1TMuon/tree/master/omtf_config

**Comment by folguera** [comment]
You need to download the cms-data PR too, it’s on the first comment.

El 10 abr 2024, a las 22:27, Zhenbin Wu ***@***.***> escribió:

﻿

It doesn't work for me. I got the missing file error:
edm::FileInPath unable to find file L1Trigger/L1TMuon/data/omtf_config/Patterns_ExtraplMB1nadMB2DTQualAndEtaFixedP_ValueP1Scale_t20_v1_SingleMu_iPt_and_OneOverPt_classProb17_recalib2_minDP0.xml anywhere in the search path.

**Comment by aloeliger** [comment]
The data files required by this PR have been added into the prototype explicitly with #1250

**Comment by aloeliger** [comment]
@folguera Given the time crunch, we're not going to spend too much time debating the nature of this here provided everything still compiles and we can run our commands. I think we can already go ahead and merge this. Please go ahead and open the PR to central CMSSW, and we'll merge this here.

**Comment by folguera** [comment]
@aloeliger @epalencia this PR requires #1196 to be merged to central before.

**Comment by aloeliger** [comment]
@folguera Can this be added to https://github.com/cms-sw/cmssw/pull/44557 then?

**Comment by aloeliger** [comment]
@folguera @zhenbinwu @epalencia @omiguelc I am going to merge this, #1250, and #1249 when I get back to my apartment in an hour or so. Are there any further expected muon fixes? Are we waiting for anything else to provide to the menu team?

**Comment by cbotta** [comment]
@folguera @zhenbinwu @omiguelc  I would like to ask if all these PRs allow the Menu team to use the GMT Standalone collection for the Standalone prompt muon efficiencies/rates, and for the Standalone displaced muon efficiencies/rates.  Thanks a lot for confirming.

@aloeliger : I think we still expect a PR to fix the GMT TkMuon efficiency at low pT

**Comment by folguera** [comment]
@cbotta YES, standalone muons, both prompt and (hopefully) displaced can be used by the menu team.

**Comment by zhenbinwu** [comment]
For this PR, the prompt OMTF looks OK. For the displaced OMTF, we need the experts to confirm the efficiency I saw. Thanks

**Comment by zhenbinwu** [comment]
Ok, please merge this PR for the SAMuon. Thanks

**Comment by zhenbinwu** [comment]
Hi, I have cherry-picked the recent PRs (https://github.com/cms-l1t-offline/cmssw/pull/1228, https://github.com/cms-l1t-offline/cmssw/pull/1239, https://github.com/cms-l1t-offline/cmssw/pull/1249, https://github.com/cms-l1t-offline/cmssw/pull/1251) to the GMT PR (https://github.com/cms-sw/cmssw/pull/44498), expect this one. 

Part of this PR is in the Phase-2 OMTF PR (https://github.com/cms-sw/cmssw/pull/44557). So I will leave these commits to that PR.

**Comment by aloeliger** [comment]
@folguera For our records, I need to understand if this has ended up in a PR to CMSSW, and where it ended up finally. Is there a PR to CMSSW with these commits in them?

**Comment by folguera** [comment]
Not yet, we need to add them to https://github.com/cms-sw/cmssw/pull/44557

**Comment by zhenbinwu** [comment]
Since the OMTF PR need to be merged before the GMT PR, I think it would be more practical to split the changes of `L1Trigger/L1TMuonOverlapPhase2/python/fakeOmtfParamsPhase2_cff.py` and `L1Trigger/L1TMuonOverlapPhase2/python/simOmtfPhase2Digis_extrapol_cfi.py` to the OMTF PR,  `L1Trigger/Configuration/python/SimL1Emulator_cff.py` and  `L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTFwdMuonTranslator.cc ` to the GMT PR. 

I can just commit the changes to the GMT branch, dropping Santi's commit history, if that is OK.

**Comment by folguera** [comment]
Sounds good

El 22 abr 2024, a las 17:09, Zhenbin Wu ***@***.***> escribió:

﻿

Since the OMTF PR need to be merged before the GMT PR, I think it would be more practical to split the changes of L1Trigger/L1TMuonOverlapPhase2/python/fakeOmtfParamsPhase2_cff.py and L1Trigger/L1TMuonOverlapPhase2/python/simOmtfPhase2Digis_extrapol_cfi.py to the OMTF PR, L1Trigger/Configuration/python/SimL1Emulator_cff.py and L1Trigger/Phase2L1GMT/plugins/Phase2L1TGMTFwdMuonTranslator.cc to the GMT PR.

I can just commit the changes to the GMT branch, dropping Santi's commit history, if that is OK.

**Comment by epalencia** [comment]
@zhenbinwu, was this fixed finally ported to CMSSW? If so, where?

**Comment by zhenbinwu** [comment]
> @zhenbinwu, was this fixed finally ported to CMSSW? If so, where?

Yes, part of this PR was added to GMT PR  https://github.com/cms-sw/cmssw/pull/44498/commits/642efa55a6aebdaec3bb5a86a0cd09850c142f9b

I think the OMTF part are merged in CMSSW as well.

**Pull request merged**